#### P452-3 T1DAL ATAC-seq DiffBind Analysis ####

## This script performs differential accessibility analysis 

#### LOAD LIBRARIES ####

library(tidyverse)
library(TxDb.Hsapiens.UCSC.hg38.knownGene) # human transcript
library(EnsDb.Hsapiens.v86)
library(stringr) # string processing
library(readxl)
library(soGGi) # for calcuation of TSSe scores
library(RColorBrewer)
library(igraph)
library(ggplot2); # plotting
# set theme
theme_set(
  theme_bw(20) +
    theme(panel.grid.major = element_blank(),
          panel.grid.minor = element_blank(),
          panel.border = element_rect(colour="black", fill=NA, size=1),
          axis.text=element_text(colour="black"),
          axis.ticks=element_line(colour="black"),
          legend.key = element_blank(),
          text = element_text(size=10),
          strip.text.x = element_text(size = 10,margin = margin( b = 2, t = 2) ),
          strip.background = element_rect(fill="white", colour="black"))) +
  scale_color_brewer(palette = "Dark2")
library(ATACseqQC)
library(GenomicAlignments)
library(DESeq2)
library(egg)
library(ComplexHeatmap)
library(DiffBind)
library(ChIPpeakAnno)
library(reactome.db) # to get enrichment paths
library(gridExtra)
library(ComplexHeatmap)
library(Repitools)
library(gridGraphics)
library(grid)
library(ggfortify)
library(gplots)
library(devtools)
#install_github("js229/Vennerable")
library(Vennerable)
library(eulerr)
library(KEGGREST)
library(gridExtra)
library(karyoploteR)
library(wiggleplotr)
library(GenomicRanges)
library(GenomicFeatures)
library(biomaRt)
library(ensembldb)


# cell pop color codes: # DN non naive= 7497d9ff, Pd1 = ba4d4cc7, CD57= 93a24eff 

# setwd
setwd("/Users/ewitkop/Box/EW_Bioinformatics_Postdoc_Research/T1DAL_10X_Project/ANALYSIS_FILES")

#### SET PATHS ####
projectname = "P452-3"
annoFileName = "P452-3 Repeat Final Annotation.xlsx"

# Paths 
baseDir <- "/Users/ewitkop/Box/EW_Bioinformatics_Postdoc_Research/T1DAL_10X_Project/ANALYSIS_FILES"
plotDir <- file.path(baseDir,'FIGURES/ATAC_REPEAT_FIGURES')
resultDir <- file.path(baseDir,'P452_3_SAVED_DATA')
test_save <- file.path(baseDir,'P452_3_SAVED_DATA')
annotationDir <- file.path(baseDir, "RAW_DATA/P452_3/")

fragmentDir <- file.path("/Volumes/Bioinformatics/pipeline/Illumina/220802_VH00126_203_AAAMVKNHV/Project_P452-3Processed_globus_220807/insertSizes")
alignmentDir <- file.path("/Volumes/Bioinformatics/pipeline/Illumina/220802_VH00126_203_AAAMVKNHV/Project_P452-3Processed_globus_220807/shiftedAlignments")
peakDir <- file.path("/Volumes/bioinformatics/pipeline/Illumina/220802_VH00126_203_AAAMVKNHV/Project_P452-3Processed_globus_220807/peaks")

annoFile <- file.path(annotationDir,annoFileName)

#### LOAD ANNO DATA AND CREATE SAMPLE SHEET ####

# Load P452-3 annotation data
load(file =  file.path(resultDir, "P452_3_anno.Rdata"))

## Read in sample sheet generated by Stephan # ignore first 18 rows
P452_3_sample_sheet <- read_csv(file.path(annotationDir,"/for analyst/02Aug2022_AAAMVKNHV_NS2000_P452-3_ATAC.csv"), skip = 19)

# load previously formatted samplesheet
samplesheet_P452_3 <- read.csv(file = file.path(resultDir, "samplesheet_P452_3.csv")) %>% dplyr::select(-X)

#### FIGURE 1B: LOAD AND PRE-PROCESS ALL ATAC DATA ####

## RUN PEAKCOUNTING PROCEDURE ON ISILON FOR INCREASED MEMORY
# USE SCRIPT P452_3_T1DAL_Count.R

# Load the data generated on isilon where the counting has been performed, but not normalized
load(file = file.path(resultDir, "ATACseqData_P452_3.RData"))

## Calculate Number of reads for each sample overlapping a peak 
# multiply the value in the Reads column by the corresponding FRiP value to
# yield the number of reads that overlap a consensus peak
info <- dba.show(ATACseqData_P452_3)
libsizes <- cbind(LibReads=info$Reads, FRiP=info$FRiP,
                  PeakReads=round(info$Reads * info$FRiP))
rownames(libsizes) <- info$ID
libsizes

## Format names for plotting
# change plot colors and legend
# change condition names for plotting
ATACseqData_P452_3_plotting <- ATACseqData_P452_3
ATACseqData_P452_3_plotting$class[DBA_CONDITION, ] <- c(
  "CD57+" ,  "PD-1+", "DN Non-naive" , "CD57+" ,  "PD-1+" , "DN Non-naive" ,
  "CD57+" ,  "PD-1+", "DN Non-naive" , "CD57+" ,  "PD-1+" , "DN Non-naive" ,
  "CD57+" ,  "PD-1+", "DN Non-naive" , "CD57+" ,  "PD-1+" , "DN Non-naive" ,
  "CD57+" ,  "PD-1+", "DN Non-naive" , "CD57+" ,  "PD-1+" , "DN Non-naive")

ATACseqData_P452_3_plotting$class[DBA_FACTOR, ] <- c(
  "0","0","0","30","30","30","0","0",
  "0","30","30","30","0","0","0","30",
  "30","30","0","0","0","30","30","30")


### PLOT FIGURE 1B
# change heatmap labels
ATACseqData_P452_3_plotting$config$condition <- "Cell Population"
ATACseqData_P452_3_plotting$config$replicate <- "Donor"
ATACseqData_P452_3_plotting$config$factor <- "Timepoint"
ATACseqData_P452_3_plotting$class[DBA_ID,] <- "" # remove row labels by setting to NULL

pdf(file.path(plotDir, "DBAHeatmap_1000_P452_3_labels.pdf"),
    height = 4,
    width = 5)
dba.plotHeatmap(ATACseqData_P452_3_plotting,  correlations=FALSE, scale="row",
                colSideCols = list(DBA_CONDITION = c("black","grey"),
                                   DBA_FACTOR = c("#93a24eff","#ba4d4cc7","#7497d9ff"),
                                   DBA_REPLICATE = c("grey20","grey40","grey60","grey80")), 
                margin = 2)
dev.off()

# NOTE: I added in the legend for Figure 1B after the fact 

# extract correlation matrix to plot with other tool
ATACseqData_P452_3_plotting_cor.matrix <- plot(ATACseqData_P452_3_plotting)

## Remove blacklisted regions

# run blacklist removal for real to see the results
ATACseqData_P452_3 <- dba.blacklist(ATACseqData_P452_3, blacklist = DBA_BLACKLIST_GRCH38, greylist = FALSE) 

# save data with the blacklist regions removed
#save(ATACseqData_P452_3, 
#     file = file.path(resultDir, "ATACseqData_P452_3_AfterNormalization_1000_blacklist_removed.RData"))

load(file= file.path(resultDir, "ATACseqData_P452_3_AfterNormalization_1000_blacklist_removed.RData"))

#### RUN DIFFERENTIAL ACCESSIBILITY ANALYSIS FOR CELL TYPE ON ALL SAMPLES  ####

## Set up contrasts to be used in every test 
ATACseqData_P452_3_no_norm <- dba.contrast(ATACseqData_P452_3, 
                                           categories=DBA_CONDITION , # Cell Type
                                           reorderMeta=list(Condition="DN_non_naive")) 
# Show contrasts
dba.show(ATACseqData_P452_3_no_norm, bContrasts=T, th = 0.05)
#Factor         Group Samples        Group2 Samples2
#1 Condition   CD57_pos_DP       8 CD57_minus_DP        8
#2 Condition   CD57_pos_DP       8  DN_non_naive        8
#3 Condition CD57_minus_DP       8  DN_non_naive        8

### 6. Running the safest option of normalizing by full library size at low computational cost (background = FALSE)

#In the absence of spike-ins or a parallel factors, the "safest" method is probably to set background=TRUE 
#and normalize=DBA_NORM_NATIVE, resulting in the use of background reads and the native normalization method
#(TMM for edgeR , and RLE for DESeq2 ). This can be approximated at very low computational cost, with no extra 
#reading of bam files, by the default settings of library=DBA_LIBSIZE_FULL, normalize=DBA_NORM_LIBRARY, 
#and back ground=FALSE.

ATACseqData_P452_3_norm <- dba.normalize(ATACseqData_P452_3_no_norm, 
                                         method=DBA_ALL_METHODS,  
                                         library=DBA_LIBSIZE_FULL, 
                                         normalize=DBA_NORM_LIB, 
                                         background=FALSE)

# get normalization factors to use for Bigwig scaling: 
ATACseqData_P452_3_norm$norm$edgeR$norm.facs # note the samples are in order from 1-24, same as the files saved on isilon
#[1] 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1

# Analyze the data
ATACseqData_P452_3_norm  <- dba.analyze(ATACseqData_P452_3_norm, method = DBA_ALL_METHODS)

# save normalized data
#save(ATACseqData_P452_3_norm, file = file.path(resultDir, "ATACseqData_P452_3_norm_final.RData"))

#### WRITE FUNCTION TO ANNOTATE PEAKS ####

# Annotate peaks using function 
# the annotatePeakInBatch function obtains the distance to the nearest TSS, 
# miRNA, exon et al for a list of peak locations leveraging IRanges and biomaRt package

peak_annotate <- function(data, list) {
  
  anno <- annotatePeakInBatch(list, 
                              AnnotationData = data, 
                              output = "overlapping",
                              bindingRegion = c(-5000, 5000),
                              select = "first", 
                              multiple =  FALSE)
  anno <- data.frame(anno) 
  anno
}

## Get anno data

## Format Annotation data using formatted annotation from Matt L/Hannah
annoData <- readRDS(file.path(baseDir,"P452_1_SAVED_DATA/annoData_human_v86.rds")) 
wantedLevels = paste("chr", c(1:22, "X", "Y"), sep = '')
noChrLevels = c(1:22, "X", "Y")

idxVec = unlist(lapply(annoData@seqnames,FUN = function(y) any(str_detect(y, paste("chr", c(1:22, "X", "Y"), sep = '')))))
annoData = annoData[idxVec]
annoData = keepSeqlevels(annoData, wantedLevels)
seqlevelsStyle(annoData) <- "UCSC" 

#### TABLE S2 SHEET 4: Analyze results of CD57pos vs CD57minus ####

load( file = file.path(resultDir, "ATACseqData_P452_3_norm_final.RData"))

## Set seqlevel style for comparisons
ATACseqData_P452_3_norm_db_anno <- dba.report(ATACseqData_P452_3_norm,
                                              method=DBA_EDGER, 
                                              contrast = 1,  
                                              bUsePval = FALSE, # tells it to use the FDR threshold 
                                              th=0.05)
# save anno before annotation
#save(ATACseqData_P452_3_norm_db_anno, file= file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57_vs_PD1.RData"))
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno   ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno <- peak_annotate(ATACseqData_P452_3_norm_db_anno, data = annoData)

# Filter for only those peaks that overlapStart and are increased - 29 sites (vs. 18 site in original sequencing) overlap start and are increased accessibility in CD57+
ATACseqData_P452_3_norm_db_anno_up_start <- ATACseqData_P452_3_norm_db_anno %>% dplyr::filter(insideFeature == "overlapStart") %>% dplyr::filter(Fold >=0)
#write.csv(ATACseqData_P452_3_norm_db_anno_up_start, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_up_start.csv"), row.names = FALSE)

# Filter for only those peaks that overlapStart and are DECREASED overlap start and are increased accessibility in CD57+
ATACseqData_P452_3_norm_db_anno_down_start <- ATACseqData_P452_3_norm_db_anno %>% dplyr::filter(insideFeature == "overlapStart") %>% dplyr::filter(Fold < 0)
#write.csv(ATACseqData_P452_3_norm_db_anno_down_start, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_down_start.csv"), row.names = FALSE)

## THIS DATA WAS USED FOR TABLE S2 SHEET 4 "CD57+DP_vs_CD57-DP_0.05"
#write.csv(ATACseqData_P452_3_norm_db_anno, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno.csv"), row.names = FALSE)
nrow(ATACseqData_P452_3_norm_db_anno) # 292

#### TABLE S2 SHEET 1: EXTRACT SIGNIFICANT results of CD57pos vs DN non naive ####

# Set seqlevel style for comparisons
ATACseqData_P452_3_norm_db_anno_CD57pos_DN <- dba.report(ATACseqData_P452_3_norm,
                                                         method=DBA_EDGER, 
                                                         contrast = 2,  # #2 Condition   CD57_pos_DP       3  DN_non_naive        8
                                                         bUsePval = FALSE, # tells it to use the FDR threshold 
                                                         th=0.05)
ATACseqData_P452_3_norm_db_anno_CD57pos_DN # 5821 ranges

# save results before annotation
#save(ATACseqData_P452_3_norm_db_anno_CD57pos_DN, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57pos_DN.Rdata"))
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno_CD57pos_DN  ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno_CD57pos_DN <- peak_annotate(ATACseqData_P452_3_norm_db_anno_CD57pos_DN, data = annoData)
#write.csv(ATACseqData_P452_3_norm_db_anno_CD57pos_DN, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57pos_DN.csv"), row.names = FALSE)

# filter only for those that are up
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_up <- ATACseqData_P452_3_norm_db_anno_CD57pos_DN %>% dplyr::filter(Fold >=0) %>%
  arrange(desc(Fold))
### THIS SHEET WAS USED FOR TABLE S2 SHEET 1 "CD57+DP_vs_DN_Up_0.05"
#write.csv(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_up, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57pos_DN_up.csv"), row.names = FALSE)

#### TABLE S2 SHEET 2: EXTRACT SIGNIFICANT results of CD57minus vs DN non naive ####

# Set seqlevel style for comparisons
ATACseqData_P452_3_norm_db_anno_CD57minus_DN <- dba.report(ATACseqData_P452_3_norm,
                                                           method=DBA_EDGER, 
                                                           contrast = 3,  # 3 Condition CD57_minus_DP       8  DN_non_naive        8
                                                           bUsePval = FALSE, # tells it to use the FDR threshold 
                                                           th=0.05)
ATACseqData_P452_3_norm_db_anno_CD57minus_DN # 3530 ranges

# save before annotating
#save(ATACseqData_P452_3_norm_db_anno_CD57minus_DN, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57minus_DN.RData"))
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno_CD57minus_DN  ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno_CD57minus_DN  <- peak_annotate(ATACseqData_P452_3_norm_db_anno_CD57minus_DN , data = annoData)
#write.csv(ATACseqData_P452_3_norm_db_anno_CD57minus_DN, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57minus_DN.csv"), row.names = FALSE)

# filter for only those that are up not just up start
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_up <- ATACseqData_P452_3_norm_db_anno_CD57minus_DN %>% dplyr::filter(Fold >=0) %>%
  arrange(desc(Fold))
### THIS SHEET WAS USED FOR TABLE S2 SHEET 2 "CD57-DP_vs_DN_Up_0.05"
#write.csv(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_up, file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57minus_DN_up.csv"), row.names = FALSE)


#### FORMAT CD57POS VS CD57NEG SIGNIFICANT SITES FOR FUTURE VOLCANO PLOTS  ####

## Get fuller list of ns and s sites by setting p value threshold to 0.5
ATACseqData_P452_3_norm_db_anno_0.5 <- dba.report(ATACseqData_P452_3_norm,
                                                  method=DBA_EDGER, 
                                                  contrast = 1,  
                                                  bUsePval = FALSE, # tells it to use the FDR threshold 
                                                  th=0.5)
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno_0.5    ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno_0.5  <- peak_annotate(ATACseqData_P452_3_norm_db_anno_0.5 , data = annoData)

# Calculate -log10FDR
ATACseqData_P452_3_norm_db_anno_0.5_volcano <- ATACseqData_P452_3_norm_db_anno_0.5 %>% mutate(neglog10FDR = -log10(FDR)) %>%
  mutate(sig = case_when(FDR <=0.05 ~ "significant",
                         FDR >0.05 ~"non-significant"))

# Create column containing names of significant differentially accessible peaks
ATACseqData_P452_3_norm_db_anno_0.5_volcano$sig_names <- NA
ATACseqData_P452_3_norm_db_anno_0.5_volcano$sig_names[ATACseqData_P452_3_norm_db_anno_0.5_volcano$sig != "non-significant"] <- 
  ATACseqData_P452_3_norm_db_anno_0.5_volcano$gene_name[ATACseqData_P452_3_norm_db_anno_0.5_volcano$sig != "non-significant"]

## join with list of sites overlapping start and up in CD57minus
ATACseqData_P452_3_norm_db_anno_0.5_volcano$CD57minus_up_start <- NA

ATACseqData_P452_3_norm_db_anno_0.5_volcano$CD57minus_up_start[ATACseqData_P452_3_norm_db_anno_0.5_volcano$gene_name %in% ATACseqData_P452_3_norm_db_anno_down_start$gene_name] <- 
  ATACseqData_P452_3_norm_db_anno_0.5_volcano$gene_name[ATACseqData_P452_3_norm_db_anno_0.5_volcano$gene_name %in% ATACseqData_P452_3_norm_db_anno_down_start$gene_name]

## Join with significantly enriched stringdb network genes
# load significant stringdb
CD57minus_up_start_stringdb_sig_network <- read_table(file = file.path(resultDir, "CD57pos_vs_CD57minus_CD57minus_up_start_stringdf_sig_interactions.tsv"))

# combine node 1 and node 2 into unique list of genes
CD57minus_up_start_stringdb_sig_network_unique <- data.frame(gene_name = unique(c(CD57minus_up_start_stringdb_sig_network$`#node1`, CD57minus_up_start_stringdb_sig_network$node2)),
                                                             CD57minus_up_start_stringdb = unique(c(CD57minus_up_start_stringdb_sig_network$`#node1`, CD57minus_up_start_stringdb_sig_network$node2)))

# join with rest of volcano plot annotations
ATACseqData_P452_3_norm_db_anno_0.5_volcano <- left_join(ATACseqData_P452_3_norm_db_anno_0.5_volcano, CD57minus_up_start_stringdb_sig_network_unique)

# download the NK cell mediated cytotoxicity pathway to join later with data
NK_mediated_cyto <- keggGet(c("hsa04650"))

# the first line is the KEGG gene entry number and the second line is the gene name
NK_mediated_cyto_genes_entry <- NK_mediated_cyto[[1]]$GENE
NK_mediated_cyto_gene_names <- as.data.frame(NK_mediated_cyto_genes_entry) %>% dplyr::filter(grepl("KO:", NK_mediated_cyto_genes_entry)) %>% 
  # separate by semi colon to get the gene symbol
  separate(NK_mediated_cyto_genes_entry, into = c("gene_name", "gene_full_name"), sep = ";") %>%
  # make copy column for joining
  mutate(hsa04650_NK_gene = gene_name)

#### EXTRACT ALL NS AND S PEAKS FROM EACH CONTRAST FOR VOLCANO PLOTS AND ADD LABELS OF INTEREST ####

### CD57pos vs CD57minus

## Get full list of ns and s sites by setting p value threshold to 1
ATACseqData_P452_3_norm_db_anno_1 <- dba.report(ATACseqData_P452_3_norm,
                                                method=DBA_EDGER, 
                                                contrast = 1,  
                                                bUsePval = FALSE, # tells it to use the FDR threshold 
                                                th=1)
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno_1   ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno_1  <- peak_annotate(ATACseqData_P452_3_norm_db_anno_1 , data = annoData)

# Calculate -log10FDR and add significance
ATACseqData_P452_3_norm_db_anno_1_volcano <- ATACseqData_P452_3_norm_db_anno_1 %>% mutate(neglog10FDR = -log10(FDR)) %>%
  mutate(sig = case_when(FDR <=0.05 ~ "significant",
                         FDR >0.05 ~"non-significant"))

# Create column containing names of significant differentially accessible peaks
ATACseqData_P452_3_norm_db_anno_1_volcano$sig_names <- NA
ATACseqData_P452_3_norm_db_anno_1_volcano$sig_names[ATACseqData_P452_3_norm_db_anno_1_volcano$sig != "non-significant"] <- 
  ATACseqData_P452_3_norm_db_anno_1_volcano$gene_name[ATACseqData_P452_3_norm_db_anno_1_volcano$sig != "non-significant"]

# Add label for those in the NK cell mediated cytotoxicity pathway
ATACseqData_P452_3_norm_db_anno_1_volcano <- left_join(ATACseqData_P452_3_norm_db_anno_1_volcano,NK_mediated_cyto_gene_names[,c("gene_name","hsa04650_NK_gene")])

### CD57pos vs DN
dba.show(ATACseqData_P452_3_no_norm, bContrasts=T, th = 0.05)
#Factor         Group Samples        Group2 Samples2
#1 Condition   CD57_pos_DP       8 CD57_minus_DP        8
#2 Condition   CD57_pos_DP       8  DN_non_naive        8
#3 Condition CD57_minus_DP       8  DN_non_naive        8

## Get full list of ns and s sites by setting p value threshold to 1
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1 <- dba.report(ATACseqData_P452_3_norm,
                                                           method=DBA_EDGER, 
                                                           contrast = 2,  
                                                           bUsePval = FALSE, # tells it to use the FDR threshold 
                                                           th=1)
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1   ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1  <- peak_annotate(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1 , data = annoData)

# Calculate -log10FDR and add significance
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano <- ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1 %>% mutate(neglog10FDR = -log10(FDR)) %>%
  mutate(sig = case_when(FDR <=0.05 ~ "significant",
                         FDR >0.05 ~"non-significant"))

# Create column containing names of significant differentially accessible peaks
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano$sig_names <- NA
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano$sig_names[ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano$sig != "non-significant"] <- 
  ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano$gene_name[ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano$sig != "non-significant"]

# Add label for those in the NK cell mediated cytotoxicity pathway
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano <- left_join(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano,NK_mediated_cyto_gene_names[,c("gene_name","hsa04650_NK_gene")])

### CD57minus vs DN

## Get full list of ns and s sites by setting p value threshold to 1
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1 <- dba.report(ATACseqData_P452_3_norm,
                                                             method=DBA_EDGER, 
                                                             contrast = 3,  
                                                             bUsePval = FALSE, # tells it to use the FDR threshold 
                                                             th=1)
seqlevelsStyle(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1   ) <-"UCSC"

# Annotate peaks using function 
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1  <- peak_annotate(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1 , data = annoData)

# Calculate -log10FDR and add significance
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano <- ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1 %>% mutate(neglog10FDR = -log10(FDR)) %>%
  mutate(sig = case_when(FDR <=0.05 ~ "Significant DA",
                         FDR >0.05 ~"Non-significant DA"))

# Create column containing names of significant differentially accessible peaks
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano$sig_names <- NA
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano$sig_names[ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano$sig != "Non-significant DA"] <- 
  ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano$gene_name[ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano$sig != "Non-significant DA"]

# Add label for those in the NK cell mediated cytotoxicity pathway
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano <- left_join(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano,NK_mediated_cyto_gene_names[,c("gene_name","hsa04650_NK_gene")])

#### FIGURE 2A: PLOT VOLCANO PLOT OF SELECT GENES IN CD57 POS VS CD57 NEG ####

# add label for those in the full NK cell mediated cytotoxicity set 
ATACseqData_P452_3_norm_db_anno_0.5_volcano_NK_KEGG <- left_join(ATACseqData_P452_3_norm_db_anno_0.5_volcano, NK_mediated_cyto_gene_names[,c("gene_name","hsa04650_NK_gene")])

# Also plot the overlapping PPI networks for up and down
# load the network
ATACseqData_P452_3_norm_db_anno_up_string_interactions <- read.table(file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_up_string_interactions.tsv"), sep = "\t", header = TRUE)
ATACseqData_P452_3_norm_db_anno_up_string_interactions_gene_set <- data.frame(gene_name = unique(c(ATACseqData_P452_3_norm_db_anno_up_string_interactions$node1, ATACseqData_P452_3_norm_db_anno_up_string_interactions$node2))) %>%
  mutate(PPI = gene_name, sig = "significant")

ATACseqData_P452_3_norm_db_anno_down_string_interactions <- read.table(file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_down_string_interactions.tsv"), sep = "\t", header = TRUE)
ATACseqData_P452_3_norm_db_anno_down_string_interactions_gene_set <- data.frame(gene_name = unique(c(ATACseqData_P452_3_norm_db_anno_down_string_interactions$node1, ATACseqData_P452_3_norm_db_anno_down_string_interactions$node2))) %>%
  mutate(PPI = gene_name,  sig = "significant")

# combine lists
ATACseqData_P452_3_norm_db_anno_up_down_string_interactions <- rbind(ATACseqData_P452_3_norm_db_anno_up_string_interactions_gene_set, 
                                                                     ATACseqData_P452_3_norm_db_anno_down_string_interactions_gene_set)

ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI <- left_join(ATACseqData_P452_3_norm_db_anno_0.5_volcano_NK_KEGG, ATACseqData_P452_3_norm_db_anno_up_down_string_interactions)


ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI <- ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI %>% 
  mutate(set = case_when(Fold >= 0 ~"up", Fold <0 ~"down")) %>%
  mutate(PPI_sig_group = paste(sig, set, sep = "_"))
ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI$PPI_sig_group <- factor(ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI$PPI_sig_group,
                                                                        levels = c("non-significant_down" ,"non-significant_up"  , "significant_down"  ,   "significant_up" ),
                                                                        labels = c("Non-Significant", "Non-Significant","Significantly Up\nin PD-1+","Significantly Up\nin CD57+"))

# Make volcano plot with labels for select genes that are enriched in PPI plot
CD57pos_vs_CD57minus_volcano_labels <- data.frame(PPI=c("NCR1","NCAM1","B3GAT1","KLRF1","KIR3DL2","KIR2DL3"))
CD57pos_vs_CD57minus_volcano_labels$PPI_subset <- CD57pos_vs_CD57minus_volcano_labels$PPI
ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_subset <- left_join(ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI,
                                                                    CD57pos_vs_CD57minus_volcano_labels)

ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_plot_subset <- 
  ggplot(ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_subset  , 
         aes(x = Fold, neglog10FDR, color = PPI_sig_group, label = PPI_subset)) + 
  geom_point(size = 2) +
  ggrepel::geom_text_repel(color = "black", size = 7, max.overlaps = 30 ) +
  #annotate("text", x = -1, y = 6.5, label = "PPI enrichment\np-value 2.74e-07", size = 3) +
  scale_color_manual(values = c("gray", "#ba4d4cc7", "#93a24eff")) +
  #theme(legend.title= element_blank(), legend.text = element_text(size = 16), text = element_text(size = 16),
  #      legend.position="bottom") +
  theme(text = element_text(size= 20), legend.position="none") +
  labs(x = "log2 Fold Change",y = "-log10 adj. p-value")

# add annotation under the columns

text_high <- textGrob("CD57+", gp=gpar(fontsize=20, fontface="bold"))
text_low <- textGrob("PD-1+", gp=gpar(fontsize=20, fontface="bold"))

ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_plot_subset <- ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_plot_subset + 
  annotation_custom(text_high,xmin=1.5,xmax=1.5,ymin=-0.5,ymax=-0.5) + 
  annotation_custom(text_low,xmin=-2,xmax=-2,ymin=-0.5,ymax=-0.5) +
  coord_cartesian( clip="off")

# FIGURE 2A: export plot
ggsave(plot = ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_plot_subset, 
       file = file.path(plotDir, "ATACseqData_P452_3_norm_db_anno_0.5_volcano_PPI_plot_subset.pdf"),device = "pdf", width = 7, height = 7)


#### FIGURE S2, TABLE S2 SHEET 3: Assess overlap of sites between the contrasts ####

## Create dbs report objects for each of the three contrasts

# Create report to save and add to full report
ATACseqData_P452_3_norm_db <- dba.report(ATACseqData_P452_3_norm, bDB=TRUE, bGain=TRUE, bLoss=TRUE,
                                         method=DBA_EDGER, 
                                         contrast = 1,  # contrast 1 = CD57_pos_DP       8 CD57_minus_DP 8
                                         bUsePval = FALSE, # tells it to use the FDR threshold 
                                         th=0.05)
ATACseqData_P452_3_norm_db$class[DBA_ID,] <- "Full_Native_EDGER"
ATACseqData_P452_3_norm_db$class[DBA_FACTOR,] <- DBA_NORM_LIB
ATACseqData_P452_3_norm_db$class[DBA_CONDITION,] <- DBA_LIBSIZE_FULL
ATACseqData_P452_3_norm_db$class[DBA_TREATMENT,] <- DBA_EDGER
dbs <- dba.peakset(ATACseqData_P452_3_norm_db)

# CD57pos vs DN
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs <- dba.report(ATACseqData_P452_3_norm, bDB=TRUE, bGain=TRUE, bLoss=TRUE,
                                                             method=DBA_EDGER, 
                                                             contrast = 2,  
                                                             bUsePval = FALSE, # tells it to use the FDR threshold 
                                                             th=0.05)
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs$class[DBA_ID,] <- "CD57pos_vs_DN"
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs$class[DBA_FACTOR,] <- DBA_NORM_NATIVE
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs$class[DBA_CONDITION,] <- DBA_LIBSIZE_FULL
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs$class[DBA_TREATMENT,] <- DBA_EDGER
dbs <- dba.peakset(dbs, ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs)
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_dbs

# CD57neg vs DN
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs <- dba.report(ATACseqData_P452_3_norm, bDB=TRUE, bGain=TRUE, bLoss=TRUE,
                                                               method=DBA_EDGER, 
                                                               contrast = 3,  
                                                               bUsePval = FALSE, # tells it to use the FDR threshold 
                                                               th=0.05)
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs$class[DBA_ID,] <- "CD57minus_vs_DN"
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs$class[DBA_FACTOR,] <- DBA_NORM_NATIVE
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs$class[DBA_CONDITION,] <- DBA_LIBSIZE_FULL
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs$class[DBA_TREATMENT,] <- DBA_EDGER
dbs <- dba.peakset(dbs, ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs)
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_dbs

### Review only overlap of CD57minus vs DN and CD57pos vs DN ###
# only look at sites with increased accessibility compared to DN - these are the genes actively coming on here! 
Venn_CD57minus_CD57pos_vsDN <- dba.plotVenn(dbs,c( 6, 9),   main="Commparison of All Read Set Contrasts",
                                            label1 =  "CD57pos_vs_DN", label2= "CD57minus_vs_DN")
# Get common peaks that are up and overlap the start site
Venn_CD57minus_CD57pos_vsDN_in_all <- Venn_CD57minus_CD57pos_vsDN$inAll
Venn_CD57minus_CD57pos_vsDN_onlyA <- Venn_CD57minus_CD57pos_vsDN$onlyA
Venn_CD57minus_CD57pos_vsDN_onlyB <- Venn_CD57minus_CD57pos_vsDN$onlyB

seqlevelsStyle(Venn_CD57minus_CD57pos_vsDN_in_all) <-"UCSC"
seqlevelsStyle(Venn_CD57minus_CD57pos_vsDN_onlyA )<- "UCSC"
seqlevelsStyle(Venn_CD57minus_CD57pos_vsDN_onlyB )<- "UCSC"

# make list to annotate
Venn_CD57minus_CD57pos_constrasts_list <-  c(
  Venn_CD57minus_CD57pos_vsDN_in_all =Venn_CD57minus_CD57pos_vsDN_in_all,
  Venn_CD57minus_CD57pos_vsDN_onlyA  =Venn_CD57minus_CD57pos_vsDN_onlyA ,
  Venn_CD57minus_CD57pos_vsDN_onlyB  =Venn_CD57minus_CD57pos_vsDN_onlyB )

# Annotate the edgeR list 
Venn_CD57minus_CD57pos_constrasts_annotate <- lapply(Venn_CD57minus_CD57pos_constrasts_list, peak_annotate, data = annoData)
names(Venn_CD57minus_CD57pos_constrasts_annotate) <-  c("Venn_CD57minus_CD57pos_vsDN_in_all",
                                                        "Venn_CD57minus_CD57pos_vsDN_onlyA",
                                                        "Venn_CD57minus_CD57pos_vsDN_onlyB")
list2env(Venn_CD57minus_CD57pos_constrasts_annotate,  envir = .GlobalEnv)

# find sites in overlaps that are also in the up_start genes lists from above
Venn_CD57minus_CD57pos_vsDN_in_all_up_start <- Venn_CD57minus_CD57pos_vsDN_in_all %>% dplyr::filter(insideFeature == "overlapStart")
nrow(Venn_CD57minus_CD57pos_vsDN_in_all_up_start) # 240
# 240 shared sites that overlap the start site and are increased


# A is CD57pos vs DN
Venn_CD57minus_CD57pos_vsDN_onlyA_up_start <- Venn_CD57minus_CD57pos_vsDN_onlyA %>% dplyr::filter(insideFeature == "overlapStart") 
nrow(Venn_CD57minus_CD57pos_vsDN_onlyA_up_start) # 322

# B is CD57minus vs DN
Venn_CD57minus_CD57pos_vsDN_onlyB_up_start <- Venn_CD57minus_CD57pos_vsDN_onlyB %>% dplyr::filter(insideFeature == "overlapStart")
nrow(Venn_CD57minus_CD57pos_vsDN_onlyB_up_start)

write.csv(Venn_CD57minus_CD57pos_vsDN_in_all_up_start, file = file.path(resultDir, "Venn_CD57minus_CD57pos_vsDN_in_all_up_start.csv"), row.names = FALSE)
write.csv(Venn_CD57minus_CD57pos_vsDN_onlyA_up_start, file = file.path(resultDir, "Venn_CD57minus_CD57pos_vsDN_onlyA_up_start.csv"), row.names = FALSE)
write.csv(Venn_CD57minus_CD57pos_vsDN_onlyB_up_start, file = file.path(resultDir, "Venn_CD57minus_CD57pos_vsDN_onlyB_up_start.csv"), row.names = FALSE)

# also export all overlapping annotated up sites

# Sheet below is TABLE S2 Sheet 3 "Shared_Up_0.05"
write.csv(Venn_CD57minus_CD57pos_vsDN_in_all, file = file.path(resultDir, "Venn_CD57minus_CD57pos_vsDN_in_all_up.csv"), row.names = FALSE)

write.csv(Venn_CD57minus_CD57pos_vsDN_onlyA, file = file.path(resultDir, "Venn_CD57minus_CD57pos_vsDN_onlyA_up.csv"), row.names = FALSE)
write.csv(Venn_CD57minus_CD57pos_vsDN_onlyB, file = file.path(resultDir, "Venn_CD57minus_CD57pos_vsDN_onlyB_up.csv"), row.names = FALSE)

## PLOT FIGURE S2 venn diagram
## plot the Venn diagram with shared up sites only
# Input in the form of a named numeric vector
fit2 <- euler(c(
  B = length(Venn_CD57minus_CD57pos_vsDN$onlyB),
  C = length(Venn_CD57minus_CD57pos_vsDN$onlyA),
  "B&C"= length(Venn_CD57minus_CD57pos_vsDN$inAll)))

pdf(file = file.path(plotDir, "P452_3_ATAC_Venn_eulerr_CD57minus_CD57pos.pdf"),
    width = 3, height = 2)
plot(fit2,
     quantities = TRUE,
     fill = "transparent",
     lty = 1:2,
     labels = c("PD-1+\nvs. DN UP", "CD57+\nvs. DN UP"))
dev.off()

#### FIGURE 1C: Volcano plot of shared exhausted genes up in CD57- and CD57+ vs DN ####

# make compiled list of exhaustion genes - removed some for clarity
ex_list <- c("GZMB", "GZMH", "CD244", "NKG7", "NCR1",  "PDCD1", "KLRF1","LAG3","SLAMF6", "TOX","TBX21","CTLA4","HAVCR2", "TIGIT","KLRG1")

## Make volcano plot of these genes using previously created objects with all genes available to plot
# also color the dots based on whether they are shared increased in both populations
ex_list_join <- data.frame(gene_name = ex_list, exhausted_list = ex_list)

# add list of annotated shared up ATAC genes
Venn_CD57minus_CD57pos_vsDN_in_all_join <- Venn_CD57minus_CD57pos_vsDN_in_all[,c('seqnames',"start","end","width","insideFeature","gene_name")] %>%
  mutate(shared = "shared")

## CD57pos_DN
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure <- left_join(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano,ex_list_join) %>%
  mutate(up_down = case_when(Fold >= 0 ~"up",
                             Fold < 0 ~"down")) %>%
  mutate(up_down_sig = paste(sig, up_down, sep = "_")) %>% 
  # add in list of shared genes
  left_join(., Venn_CD57minus_CD57pos_vsDN_in_all_join) %>%
  mutate(shared = replace_na(shared, "not_shared")) %>%
  mutate(up_down_sig_shared = paste(sig, up_down, shared, sep = "_")) %>% 
  mutate(up_down_sig_shared = case_when(up_down_sig_shared =="significant_down_not_shared" ~"DN",
                                        up_down_sig_shared =="significant_up_shared" ~ "Tex Shared UP",
                                        up_down_sig_shared =="significant_up_not_shared" ~ "CD57+",
                                        up_down_sig_shared =="non-significant_up_not_shared" ~ "NS",
                                        up_down_sig_shared =="non-significant_down_not_shared" ~ "NS"
  ))


# get rows to remove that have a duplicated label by removing the one with the lower fold change
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure_rm <- ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure %>%
  dplyr::filter(!is.na(exhausted_list)) %>% 
  group_by(exhausted_list) %>%
  dplyr::filter(n() >1) %>%
  top_n(-1, abs(Fold))

# use anti-join to remove
ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure <- anti_join(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure,ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure_rm)

# Make volcano plot labelled with genes of interest for paper and shared genes in both contrasts
CD57pos_DN_volcano_exhausted <- 
  ggplot(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure,  aes(x = Fold, neglog10FDR, color = up_down_sig_shared,  label = exhausted_list)) + 
  geom_point() +
  ggrepel::geom_text_repel(data = subset(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure, exhausted_list %in% c("SLAMF6","TBX21","TOX","LAG3","CTLA4","PDCD1", "HAVCR2", "TIGIT","KLRG1","CD244")),
                           color = "black", fontface = "bold", size = 5, max.overlaps = 20, box.padding = 0.25, nudge_x = 0.2 ) +
  ggrepel::geom_text_repel(
    data = subset(ATACseqData_P452_3_norm_db_anno_CD57pos_DN_1_volcano_figure, !exhausted_list %in% c("SLAMF6","TBX21","TOX","LAG3","CTLA4","PDCD1", "HAVCR2", "TIGIT","KLRG1","CD244")),
    color = "black", size = 5, fontface = "italic", min.segment.length = unit(0, 'lines'), max.overlaps = 20 ,box.padding = 0.25 ) +
  # scale_color_manual(values = c("gray","black","#93a24eff", "#7497d9ff")) + 
  scale_color_manual(values = c("#93a24eff", "#7497d9ff","gray","#9350a1"))+ 
  #theme(legend.title= element_blank(), legend.text = element_text(size = 20), text = element_text(size = 20),
  theme(text = element_text(size= 20),legend.title= element_blank(), legend.position = "bottom", legend.text = element_text(size = 16),
        title = element_text(size = 16)) +
  #theme(text = element_text(size= 20)) +
  scale_y_continuous(limits = c(0,13), breaks = c(0,5,10,13))+
  labs(x = "log2 Fold Change",y = "-log10 adj. p-value", title = "Shared Cytoxicity and Exhaustion Genes",
       colour = NULL) +
  guides(colour = guide_legend(override.aes = list(size=8))) # increase size of legend dots

# add annotation under the columns

text_high <- textGrob("CD57+", gp=gpar(fontsize=20, fontface="bold"))
text_low <- textGrob("DN", gp=gpar(fontsize=20, fontface="bold"))

CD57pos_DN_volcano_exhausted_plot <- CD57pos_DN_volcano_exhausted + 
  annotation_custom(text_high,xmin=3.5,xmax=4,ymin=-3,ymax=-3) + 
  annotation_custom(text_low,xmin=-3.8,xmax=-3.8,ymin=-2.5,ymax=-2.5) +
  coord_cartesian( clip="off")

# export plot for FIGURE 1C
ggsave(plot = CD57pos_DN_volcano_exhausted_plot, 
       file = file.path(plotDir, "CD57pos_DN_volcano_exhausted_plot.pdf"),device = "pdf",  width = 7.2, height = 6)


## Repeat for CD57minus vs DN
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure <- left_join(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano,ex_list_join) %>%
  mutate(up_down = case_when(Fold >= 0 ~"up",
                             Fold < 0 ~"down")) %>%
  mutate(up_down_sig = paste(sig, up_down, sep = "_")) %>% 
  # add in list of shared genes
  left_join(., Venn_CD57minus_CD57pos_vsDN_in_all_join) %>%
  mutate(shared = replace_na(shared, "not_shared")) %>%
  mutate(up_down_sig_shared = paste(sig, up_down, shared, sep = "_")) %>% 
  mutate(up_down_sig_shared = case_when(up_down_sig_shared =="Significant DA_down_not_shared" ~"DN",
                                        up_down_sig_shared =="Significant DA_up_shared" ~ "Tex Shared UP",
                                        up_down_sig_shared =="Significant DA_up_not_shared" ~ "PD-1+",
                                        up_down_sig_shared =="Non-significant DA_up_not_shared" ~ "NS",
                                        up_down_sig_shared =="Non-significant DA_down_not_shared" ~ "NS"
  ))

# get rows to remove that have a duplicated label by removing the one with the lower fold change
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure_rm <- ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure %>%
  dplyr::filter(!is.na(exhausted_list)) %>% 
  group_by(exhausted_list) %>%
  dplyr::filter(n() >1) %>%
  top_n(-1, abs(Fold))

# use anti-join to remove
ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure <- anti_join(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure,ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure_rm)

# Make volcano plot with labeled only those that are in the NK cell mediated cytotoxicity pathway
CD57minus_DN_volcano_exhausted <- ggplot(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure  , 
                                         aes(x = Fold, neglog10FDR, color = up_down_sig_shared,  label = exhausted_list)) + 
  geom_point() +
  ggrepel::geom_text_repel(data = subset(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure, exhausted_list %in% c("SLAMF6","TBX21","TOX","LAG3","CTLA4","PDCD1", "HAVCR2", "TIGIT","KLRG1","CD244")),
                           color = "black", fontface = "bold", size = 4.5, max.overlaps = 20, nudge_y =0.2, nudge_x = 0.35,  min.segment.length = unit(0.1, "lines") ) +
  ggrepel::geom_text_repel(
    data = subset(ATACseqData_P452_3_norm_db_anno_CD57minus_DN_1_volcano_figure, !exhausted_list %in% c("SLAMF6","TBX21","TOX","LAG3","CTLA4","PDCD1", "HAVCR2", "TIGIT","KLRG1","CD244")),
    color = "black", size = 4.5, fontface = "italic", max.overlaps = 20, nudge_x = -0.2, nudge_y = 0.5) +
  # scale_color_manual(values = c("gray","black","#93a24eff", "#7497d9ff")) + 
  scale_color_manual(values = c(  "#7497d9ff","gray","#ba4d4cc7","#9350a1"))+ 
  #theme(legend.title= element_blank(), legend.text = element_text(size = 20), text = element_text(size = 20),
  theme(text = element_text(size= 20),legend.title= element_blank(), legend.position = "bottom", legend.text = element_text(size = 16),
        title = element_text(size = 16)) +
  scale_y_continuous(limits = c(0,13), breaks = c(0,5,10,13))+
  scale_x_continuous(limits = c(-3,5), breaks = c(-2,0,2,4)) +
  labs(x = "log2 Fold Change",y = "-log10 adj. p-value", title = "Shared Cytoxicity and Exhaustion Genes",
       colour = NULL)+
  guides(colour = guide_legend(override.aes = list(size=8)))

# add annotation under the columns

text_high <- textGrob("PD-1+", gp=gpar(fontsize=20, fontface="bold"))
text_low <- textGrob("DN", gp=gpar(fontsize=20, fontface="bold"))

CD57minus_DN_volcano_exhausted_plot <- CD57minus_DN_volcano_exhausted + 
  annotation_custom(text_high,xmin=4,xmax=4,ymin=-3,ymax=-3) + 
  annotation_custom(text_low,xmin=-3,xmax=-3,ymin=-2.5,ymax=-2.5) +
  coord_cartesian( clip="off")

# export plot for FIGURE 1C
ggsave(plot = CD57minus_DN_volcano_exhausted_plot, 
       file = file.path(plotDir, "CD57minus_DN_volcano_exhausted_plot.pdf"),device = "pdf", width = 7.2, height = 6)


#### FIGURE 1D Bubble plot OF pathway enrichment in CD57pos vs DN and CD57neg vs DN ####

### Recreate bubble plot for all up sites and not just up start

# from files ATACseqData_P452_3_norm_db_anno_CD57minus_DN_up.csv, ATACseqData_P452_3_norm_db_anno_CD57pos_DN_up.csv
#remove the # from beginning of file first
CD57pos_DN_up_enrichment.KEGG <- read.table(file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57pos_DN_up_enrichment.KEGG.tsv"), sep = "\t", header = TRUE)
CD57pos_DN_up_enrichment.KEGG$group <- "CD57pos_DN_up"
CD57minus_DN_up_enrichment.KEGG <- read.table(file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_CD57minus_DN_up_enrichment.KEGG.tsv"), sep = "\t", header = TRUE)
CD57minus_DN_up_enrichment.KEGG$group <- "CD57minus_DN_up"

# combine the number of observed genes and FDR for each contrast into table
CD57pos_CD57minus_DN_up_enrichment.KEGG <- rbind(CD57pos_DN_up_enrichment.KEGG, CD57minus_DN_up_enrichment.KEGG) %>%
  # remove everything greater than 0.01
  dplyr::filter(false.discovery.rate <0.01)
CD57pos_CD57minus_DN_up_enrichment.KEGG$group <- factor(CD57pos_CD57minus_DN_up_enrichment.KEGG$group, 
                                                        levels = c("CD57minus_DN_up", "CD57pos_DN_up" ),
                                                        labels = c("PD-1+ vs.\nDN UP", "CD57+ vs.\nDN UP"))


# Keep only terms that are shared between the contrasts
CD57pos_CD57minus_DN_up_enrichment.KEGG_shared <- CD57pos_CD57minus_DN_up_enrichment.KEGG  %>% group_by(term.description) %>%
  dplyr::filter(n() == 2)

# order the levels for term
CD57pos_CD57minus_DN_up_enrichment.KEGG_shared$term.description <- factor(CD57pos_CD57minus_DN_up_enrichment.KEGG_shared$term.description,
                                                                          levels = c(
                                                                            "Human cytomegalovirus infection",
                                                                            "Chemokine signaling pathway" , 
                                                                            "Allograft rejection",
                                                                            "Viral protein interaction with cytokine and cytokine receptor",
                                                                            "Natural killer cell mediated cytotoxicity",
                                                                            "Cytokine-cytokine receptor interaction"),
                                                                          labels= c("Human cytomegalovirus\ninfection",
                                                                                    "Chemokine signaling\npathway" , 
                                                                                    "Allograft\nrejection",
                                                                                    "Viral protein\ninteraction with cytokine\nand cytokine receptor",
                                                                                    "Natural killer cell\nmediated cytotoxicity",
                                                                                    "Cytokine-cytokine\nreceptor interaction"))  
## PLOT FIGURE 1D
shared_KEGG_up_bubble_only_shared <- ggplot(CD57pos_CD57minus_DN_up_enrichment.KEGG_shared, 
                                            aes(x = group, y = term.description, size =observed.gene.count, color = false.discovery.rate)) +
  geom_point() + theme( text = element_text(size = 20)) + labs(x = NULL, y = "KEGG Term", title = "Shared KEGG\nPathway Enrichment") + 
  scale_color_viridis_c(direction = -1) + scale_size_continuous(name = "Observed Gene\nCount", range = c(3,10)) +
  labs(color = "FDR p-value") 

ggsave(shared_KEGG_up_bubble_only_shared, file = file.path(plotDir, "shared_KEGG_up_bubble_only_shared.pdf"), device = "pdf", width = 8, height = 5.5)

#### RUN DIFFERENTIAL ACCESSIBILITY ANALYSIS BY TIMEPOINT FOR ALL SAMPLES ####

### CD57pos

# subset samples
ATACseqData_P452_3_CD57pos <- dba(ATACseqData_P452_3, mask = ATACseqData_P452_3$masks$CD57_pos_DP)

# get contrasts
ATACseqData_P452_3_CD57pos <- dba.contrast(ATACseqData_P452_3_CD57pos , 
                                           categories=DBA_FACTOR , # TIMEPOINT
                                           reorderMeta=list(Factor="0")) # SET TIMEPOINT 0 FIRST

### 6. Running the safest option of normalizing by full library size at low computational cost (background = FALSE)
#In the absence of spike-ins or a parallel factors, the "safest" method is probably to set background=TRUE 
#and normalize=DBA_NORM_NATIVE, resulting in the use of background reads and the native normalization method
#(TMM for edgeR , and RLE for DESeq2 ). This can be approximated at very low computational cost, with no extra 
#reading of bam files, by the default settings of library=DBA_LIBSIZE_FULL, normalize=DBA_NORM_LIBRARY, 
#and back ground=FALSE.

ATACseqData_P452_3_CD57pos_norm <- dba.normalize(ATACseqData_P452_3_CD57pos, 
                                                 method=DBA_ALL_METHODS,  
                                                 library=DBA_LIBSIZE_FULL, 
                                                 normalize=DBA_NORM_LIB, 
                                                 background=FALSE)

# Analyze the data
ATACseqData_P452_3_CD57pos_norm  <- dba.analyze(ATACseqData_P452_3_CD57pos_norm, method = DBA_ALL_METHODS)


## Show summary of contrast results

dba.show(ATACseqData_P452_3_CD57pos_norm , bContrasts=T, th = 0.05) # set FDR threshold to be 0.05
#Factor   Group Samples   Group2 Samples2 DB.edgeR DB.DESeq2
#1 Factor 30       4 0        4        0         0

# no significant between CD57 pos timepoints

#### CD57neg

# subset samples
ATACseqData_P452_3_CD57neg <- dba(ATACseqData_P452_3, mask = ATACseqData_P452_3$masks$CD57_minus_DP)

# get contrasts
ATACseqData_P452_3_CD57neg <- dba.contrast(ATACseqData_P452_3_CD57neg , 
                                           categories=DBA_FACTOR , # TIMEPOINT
                                           reorderMeta=list(Factor="0")) # SET TIMEPOINT 0 FIRST

### 6. Running the safest option of normalizing by full library size at low computational cost (background = FALSE)
#In the absence of spike-ins or a parallel factors, the "safest" method is probably to set background=TRUE 
#and normalize=DBA_NORM_NATIVE, resulting in the use of background reads and the native normalization method
#(TMM for edgeR , and RLE for DESeq2 ). This can be approximated at very low computational cost, with no extra 
#reading of bam files, by the default settings of library=DBA_LIBSIZE_FULL, normalize=DBA_NORM_LIBRARY, 
#and back ground=FALSE.

ATACseqData_P452_3_CD57neg_norm <- dba.normalize(ATACseqData_P452_3_CD57neg, 
                                                 method=DBA_ALL_METHODS,  
                                                 library=DBA_LIBSIZE_FULL, 
                                                 normalize=DBA_NORM_LIB, 
                                                 background=FALSE)

# Analyze the data
ATACseqData_P452_3_CD57neg_norm  <- dba.analyze(ATACseqData_P452_3_CD57neg_norm, method = DBA_ALL_METHODS)


## Show summary of contrast results

dba.show(ATACseqData_P452_3_CD57neg_norm , bContrasts=T, th = 0.05) # set FDR threshold to be 0.05
#Factor   Group Samples   Group2 Samples2 DB.edgeR DB.DESeq2
#1 Factor 30       4 0        4        0         0

# no significant between CD57 neg timepoint

### DN populations

# subset samples
ATACseqData_P452_3_DN_non_naive <- dba(ATACseqData_P452_3, mask = ATACseqData_P452_3$masks$DN_non_naive)

# get contrasts
ATACseqData_P452_3_DN_non_naive <- dba.contrast(ATACseqData_P452_3_DN_non_naive , 
                                                categories=DBA_FACTOR , # TIMEPOINT
                                                reorderMeta=list(Factor="0")) #  SET TIMEPOINT 0 FIRST

### 6. Running the safest option of normalizing by full library size at low computational cost (background = FALSE)
#In the absence of spike-ins or a parallel factors, the "safest" method is probably to set background=TRUE 
#and normalize=DBA_NORM_NATIVE, resulting in the use of background reads and the native normalization method
#(TMM for edgeR , and RLE for DESeq2 ). This can be approximated at very low computational cost, with no extra 
#reading of bam files, by the default settings of library=DBA_LIBSIZE_FULL, normalize=DBA_NORM_LIBRARY, 
#and back ground=FALSE.

ATACseqData_P452_3_DN_non_naive_norm <- dba.normalize(ATACseqData_P452_3_DN_non_naive, 
                                                      method=DBA_ALL_METHODS,  
                                                      library=DBA_LIBSIZE_FULL, 
                                                      normalize=DBA_NORM_LIB, 
                                                      background=FALSE)

# Analyze the data
ATACseqData_P452_3_DN_non_naive_norm  <- dba.analyze(ATACseqData_P452_3_DN_non_naive_norm, method = DBA_ALL_METHODS)


## Show summary of contrast results

dba.show(ATACseqData_P452_3_DN_non_naive_norm , bContrasts=T, th = 0.05) # set FDR threshold to be 0.05
#Factor   Group Samples   Group2 Samples2 DB.edgeR DB.DESeq2
#1 Factor 30       4 0        4        0         0
dba.show(ATACseqData_P452_3_DN_non_naive_norm , bContrasts=T, th = 0.1)
# none significant

#### FIGURE 2C: Plot Mean ATAC peaks for KIR2DL3, KIR3DL2, NCR1, and KLRF1 for all individual files using wiggleplotR ####

# make sure you're connected to the bioinformatics server first to access the bigwig files

# create sample metadata for first three samples (one sample per condition for example)
peak_plot_sample_data = dplyr::data_frame(
  sample_id = samplesheet_P452_3$Sample_ID, 
  condition = factor(samplesheet_P452_3$Condition, 
                     levels = c("CD57_pos_DP", "CD57_minus_DP",  "DN_non_naive"),
                     labels = c("CD57+", "PD-1+",  "DN")), 
  scaling_factor = 1, 
  bigWig = paste0("/Volumes/Bioinformatics/pipeline/Illumina/220802_VH00126_203_AAAMVKNHV/Project_P452-3Processed_globus_220807/shiftedAlignments/bigWigs/", samplesheet_P452_3$libId, "_scaled.bigWig"))


track_data = dplyr::mutate(peak_plot_sample_data, track_id = condition, colour_group = condition)

# extract exon and cds information as Granges objects
edb <- EnsDb.Hsapiens.v86 
exons = exonsBy(edb, by = "tx",use.names = TRUE)
cdss = cdsBy(edb,by = "tx",use.names = TRUE)
txlength = transcriptLengths(edb)

### Plot NCR1

# extract metadata for gene
ncr1_metadata <- transcripts(edb,  columns = c("seq_name","gene_id", "tx_id", "gene_name", "seq_strand", "tx_biotype", "tx_start","tx_end"),
                             filter = GeneNameFilter("NCR1", condition = "startsWith"),
                             return.type = "data.frame")
ncr1_metadata <- ncr1_metadata %>%  dplyr:: filter(tx_biotype == "protein_coding" & seq_name == 19) %>%
  dplyr::rename(transcript_id = tx_id, 
                chr=seq_name, 
                strand = seq_strand) %>% dplyr::select(-chr, -tx_biotype) %>% as_tibble()
summary(ncr1_metadata)

# extract exon info for genes of interest
ncr1_exons_ids <- ncr1_metadata$transcript_id
ncr1_exons <- exons[ncr1_exons_ids]
ncr1_cds <-  cdss[ncr1_exons_ids]

names(ncr1_exons)
names(ncr1_cds)

# find the longest transcript for the ncr1 gene
ncr1_lengths <- txlength[txlength$tx_id %in%  ncr1_exons_ids,]
ncr1_length_top <- ncr1_lengths  %>% top_n(1, tx_len)

# plotting function explicitly sets FPM as y-axis label, change this in place in the makeCoveragePlot function 
trace('makeCoveragePlot', edit = T, where = asNamespace("wiggleplotr"))
# set explicitly to "Mean Normalized Counts" and add to theme(text = element_text(size = 50)), scale_y_continuous(limits = c(0, 100)), strip.text.y = element_blank())
trace('plotTranscriptStructure', edit = T, where = asNamespace("wiggleplotr"))
# add text = element_text(size = 45) to the theme object for the plot, change transcript label size to 10, strip.text.y = element_blank()),
# strip background to element_blank also, xlab("")
# change transcript plot colors to be gray and black, change axis title to distance from start (bp), changed transcript_label 
#trace('plotCoverage', edit = T, where = asNamespace("wiggleplotr"))

# only need to run this once per session

# generate mean coverage plot using wiggplotR
ncr1_mean_coverage <- plotCoverage(ncr1_exons[ncr1_length_top$tx_id], ncr1_cds[ncr1_length_top$tx_id], ncr1_metadata, track_data =track_data,
                                   flanking_length = c(3000, 50),
                                   mean_only = TRUE, 
                                   transcript_label = FALSE,
                                   heights = c(0.8, 0.2),
                                   # # DN non naive= 7497d9ff, Pd1 = ba4d4cc7, CD57= 93a24eff 
                                   fill_palette = c("#93a24eff" , "#ba4d4cc7", "#7497d9ff"))


### Plot KIR2DL3

# extract metadata for gene
KIR2DL3_metadata <- transcripts(edb,  columns = c("seq_name","gene_id", "tx_id", "gene_name", "seq_strand", "tx_biotype"),
                                filter = GeneNameFilter("KIR2DL3", condition = "startsWith"),
                                return.type = "data.frame")
KIR2DL3_metadata <- KIR2DL3_metadata %>%   dplyr::filter(tx_biotype == "protein_coding" & seq_name == 19) %>%
  dplyr::rename(transcript_id = tx_id, 
                chr=seq_name, 
                strand = seq_strand) %>% dplyr::select(-chr, -tx_biotype) %>% as_tibble()
summary(KIR2DL3_metadata)

# extract exon info for genes of interest
KIR2DL3_exons_ids <- KIR2DL3_metadata$transcript_id
KIR2DL3_exons <- exons[KIR2DL3_exons_ids]
KIR2DL3_cds <-  cdss[KIR2DL3_exons_ids]

names(KIR2DL3_exons)
names(KIR2DL3_cds)

# find the longest transcript for the ncr1 gene
KIR2DL3_lengths <- txlength[txlength$tx_id %in%  KIR2DL3_exons_ids,]
KIR2DL3_length_top <- KIR2DL3_lengths   %>% top_n(1, tx_len)

# generate mean coverage plot using wiggplotR
KIR2DL3_mean_coverage <- plotCoverage(KIR2DL3_exons[KIR2DL3_length_top$tx_id], KIR2DL3_cds[KIR2DL3_length_top$tx_id], KIR2DL3_metadata, track_data =track_data,
                                      flanking_length = c(3000, 50),
                                      mean_only = TRUE, 
                                      transcript_label = FALSE,
                                      heights = c(0.8, 0.2),
                                      # # DN non naive= 7497d9ff, Pd1 = ba4d4cc7, CD57= 93a24eff 
                                      fill_palette = c("#93a24eff" , "#ba4d4cc7", "#7497d9ff"))

### Plot KIR3DL2

# extract metadata for gene
KIR3DL2_metadata <- transcripts(edb,  columns = c("seq_name","gene_id", "tx_id", "gene_name", "seq_strand", "tx_biotype"),
                                filter = GeneNameFilter("KIR3DL2", condition = "startsWith"),
                                return.type = "data.frame")
KIR3DL2_metadata <- KIR3DL2_metadata %>%   dplyr::filter(tx_biotype == "protein_coding" & seq_name == 19) %>%
  dplyr::rename(transcript_id = tx_id, 
                chr=seq_name, 
                strand = seq_strand) %>% dplyr::select(-chr, -tx_biotype) %>% as_tibble()
summary(KIR3DL2_metadata)

# extract exon info for genes of interest
KIR3DL2_exons_ids <- KIR3DL2_metadata$transcript_id
KIR3DL2_exons <- exons[KIR3DL2_exons_ids]
KIR3DL2_cds <-  cdss[KIR3DL2_exons_ids]

names(KIR3DL2_exons)
names(KIR3DL2_cds)

# find the longest transcript for the ncr1 gene
KIR3DL2_lengths <- txlength[txlength$tx_id %in%  KIR3DL2_exons_ids,]
KIR3DL2_length_top <- KIR3DL2_lengths   %>% top_n(1, tx_len)

# generate mean coverage plot using wiggplotR
KIR3DL2_mean_coverage <- plotCoverage(KIR3DL2_exons[KIR3DL2_length_top$tx_id], KIR3DL2_cds[KIR3DL2_length_top$tx_id], KIR3DL2_metadata, track_data =track_data,
                                      flanking_length = c(3000, 50),
                                      mean_only = TRUE, 
                                      transcript_label = FALSE,
                                      heights = c(0.8, 0.2),
                                      # # DN non naive= 7497d9ff, Pd1 = ba4d4cc7, CD57= 93a24eff 
                                      fill_palette = c("#93a24eff" , "#ba4d4cc7", "#7497d9ff"))

## Plot KLRF1


# extract metadata for gene
KLRF1_metadata <- transcripts(edb,  columns = c("seq_name","gene_id", "tx_id", "gene_name", "seq_strand", "tx_biotype"),
                              filter = GeneNameFilter("KLRF1", condition = "startsWith"),
                              return.type = "data.frame")
KLRF1_metadata <- KLRF1_metadata %>%   dplyr::filter(tx_biotype == "protein_coding" & seq_name == 12) %>%
  dplyr::rename(transcript_id = tx_id, 
                chr=seq_name, 
                strand = seq_strand) %>% dplyr::select(-chr, -tx_biotype) %>% as_tibble()
summary(KLRF1_metadata)

# extract exon info for genes of interest
KLRF1_exons_ids <- KLRF1_metadata$transcript_id
KLRF1_exons <- exons[KLRF1_exons_ids]
KLRF1_cds <-  cdss[KLRF1_exons_ids]

# find the longest transcript for the ncr1 gene
KLRF1_lengths <- txlength[txlength$tx_id %in%  KLRF1_exons_ids,]
KLRF1_length_top <- KLRF1_lengths   %>% top_n(1, tx_len)

# generate mean coverage plot using wiggplotR
KLRF1_mean_coverage <- plotCoverage(KLRF1_exons[KLRF1_length_top$tx_id], KLRF1_cds[KLRF1_length_top$tx_id], KLRF1_metadata, track_data =track_data,
                                    flanking_length = c(3000, 50),
                                    mean_only = TRUE, 
                                    transcript_label = FALSE,
                                    heights = c(0.8, 0.2),
                                    # # DN non naive= 7497d9ff, Pd1 = ba4d4cc7, CD57= 93a24eff 
                                    fill_palette = c("#93a24eff" , "#ba4d4cc7", "#7497d9ff"))


### combine and export all three plots for FIGURE 2C
pdf(file = file.path(plotDir, "mean_normalized_atac_peaks.pdf"), height = 18, width = 37.5 )
cowplot::plot_grid(ncr1_mean_coverage,KLRF1_mean_coverage ,KIR3DL2_mean_coverage, KIR2DL3_mean_coverage,
                   labels = c("NCR1","KLRF1", "KIR3DL2","KIR2DL3"),
                   label_size = 45,
                   ncol = 4, nrow = 1, align = "hv",
                   label_x = 0.1,
                   label_y = 1) +
  theme(plot.margin = unit(c(1, 1, 1, 1), "cm"))
dev.off()

#### FIGURE 2B, FIGURE S3: Make igraph connected network in CD57pos vs CD57minus up and down ####

### Load networks 
# from the file ATACseqData_P452_3_norm_db_anno_up_string_interactions.tsv 
ATACseqData_P452_3_norm_db_anno_up_string_interactions <- read.table(file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_up_string_interactions.tsv"), sep = "\t", header = TRUE)
ATACseqData_P452_3_norm_db_anno_down_string_interactions <- read.table(file = file.path(resultDir, "ATACseqData_P452_3_norm_db_anno_down_string_interactions.tsv"), sep = "\t", header = TRUE)


# create edge df 
ATACseqData_P452_3_norm_db_anno_up_string_interactions_edge <- ATACseqData_P452_3_norm_db_anno_up_string_interactions %>%
  dplyr::select(node1,node2, combined_score) %>%  
  # remove reciprocal edges
  mutate(key = paste0(pmin(node1, node2), pmax(node1, node2), sep = "")) %>% 
  distinct( key, .keep_all = TRUE)

ATACseqData_P452_3_norm_db_anno_down_string_interactions_edge <- ATACseqData_P452_3_norm_db_anno_down_string_interactions %>%
  dplyr::select(node1,node2, combined_score) %>%  
  # remove reciprocal edges
  mutate(key = paste0(pmin(node1, node2), pmax(node1, node2), sep = "")) %>% 
  distinct( key, .keep_all = TRUE)

# Create node files
ATACseqData_P452_3_norm_db_anno_up_string_interactions_nodes <- 
  data.frame(node = unique(c(ATACseqData_P452_3_norm_db_anno_up_string_interactions$node1, ATACseqData_P452_3_norm_db_anno_up_string_interactions$node2)))

ATACseqData_P452_3_norm_db_anno_down_string_interactions_nodes <- 
  data.frame(node = unique(c(ATACseqData_P452_3_norm_db_anno_down_string_interactions$node1, ATACseqData_P452_3_norm_db_anno_down_string_interactions$node2)))

# create network
venn_CD57pos_CD57minus_up <- graph_from_data_frame(d=ATACseqData_P452_3_norm_db_anno_up_string_interactions_edge , 
                                                   vertices=ATACseqData_P452_3_norm_db_anno_up_string_interactions_nodes, directed=F) 

venn_CD57pos_CD57minus_down <- graph_from_data_frame(d=ATACseqData_P452_3_norm_db_anno_down_string_interactions_edge , 
                                                     vertices=ATACseqData_P452_3_norm_db_anno_down_string_interactions_nodes, directed=F) 

# plot networks
# view igraph plotting parameters
#?igraph.plotting
# Set node size
V(venn_CD57pos_CD57minus_up)$size <- 7
V(venn_CD57pos_CD57minus_down)$size <- 7

# Set edge width based on weight:
E(venn_CD57pos_CD57minus_up)$width <- E(venn_CD57pos_CD57minus_up)$combined_score*5
E(venn_CD57pos_CD57minus_down)$width <- E(venn_CD57pos_CD57minus_down)$combined_score*5

# Set edge color 
#E(upset_pairwise_net)$color = E(upset_pairwise_net)$edge_color

# plot FIGURE 2B
pdf(file = file.path(plotDir, "venn_CD57pos_CD57minus_up_network.pdf"), width = 5, height =5 )
plot(venn_CD57pos_CD57minus_up, main = "Significant PPI Network Increased in CD57+\nP = 2.7e-07",  
     label.font  = 2, vertex.label.color="black", vertex.label.cex = 1, 
     vertex.label.dist=2,
     vertex.color = "#93a24eff")
dev.off()

## plot FIGURE S3
pdf(file = file.path(plotDir, "venn_CD57pos_CD57minus_down_network.pdf"), width = 6, height = 6 )
plot(venn_CD57pos_CD57minus_down, main = "Significant PPI Network Increased in PD-1+\nP = 4.4e-05",  
     label.font  = 2, vertex.label.color="black", vertex.label.cex = 0.8, 
     vertex.label.dist=2,
     vertex.color = "#ba4d4cc7")
dev.off()




